text/3396-extern-types-v2.md

```diff
@@ -0,0 +1,237 @@
+- Feature Name: `extern_types_v2`
+- Start Date: 2023-02-19
+- RFC PR: [rust-lang/rfcs#3396](https://github.com/rust-lang/rfcs/pull/3396)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+Define types external to Rust and introduce syntax to declare them.
+This additionally introduces the `MetaSized` trait to allow these types to be interacted with in a generic context.
+This supersedes [RFC 1861].
+
+
+# Motivation
+[motivation]: #motivation
+
+The motivation from [RFC 1861] for why we want extern types at all still applies.
+
+The summary of that is that when writing FFI bindings, the foreign code may give you a pointer to a type but not provide the layout for that type.
+The current approach for representing these types in Rust, suggested by the 'nomicon, is with a ZST like the following.
+```rust
+#[repr(C)]
+pub struct Opaque {
+    _data: [u32; 0],
+    _marker: PhantomData<(*mut u8, PhantomPinned)>
+}
+```
+The type is `repr(C)` so that it can be used in FFI.
+The `_data` field provides the alignment of the type, and the `_marker` field causes the correct set of automatically implemented traits to be implemented, in this case `!Send`, `!Sync`, `!Unpin`, but `Freeze` (`UnsafeCell` can be used to remove the last one of those).
+This works, however the compiler believes this type is statically sized and aligned which may not be true.
+This means that the type can be easily misused by placing it on the stack, allocating it on the heap, calling `ptr::read`, calling `mem::size_of`, etc.
+This proposal introduces extern types, a way to accurately represent these opaque types in the Rust type system such that the compiler prevents you from misusing it.
+
+The motivation for replacing [RFC 1861] is around computing the offset of fields in aggregate types.
+Extern types do not have an alignment known to the Rust compiler and therefore cannot be included as a field in an aggregate type as it is impossible to correctly calculate their offset, however that RFC did not address this.
+This implies that extern types cannot be included in any (non-`repr(transparent)`) structs and so must be prevented.
+
+Extern types should be able to be used in generic contexts so that they receive blanket trait implementations and can be used inside generic wrapper types.
+Prior to this, all generic types could be included as fields of structs and it was possible to obtain the size and alignment of any type.
+Extern types do not have a size or alignment so this RFC allows users to specify generic bounds sufficiently to statically prevent the above issues.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing bindings to foreign code, any types that are opaque to the Rust type system should be declared as follows:
+```rust
+extern {
+    type Foo;
+}
```

Comment:
Is the `extern type Foo` shorthand still allowed under this proposal or do all extern types have to be inside an extern block?

Similarly, how do extern blocks interact with ABI tags? For example, if I have an `extern "C"` block, is a type declaration inside it invalid, or just a warning? I would say that adding an ABI to these should be a deny-by-default lint that states that the tag is unused, with the potential for them having some meaning in the future.

Reply:
Note that `extern { type Blah; }` will currently be reformatted by rustfmt to `extern "C" { type Blah; }`. So this question should probably have t-style weigh in.

Reply:
This question gets at one of the niggles I have with this proposal. The ABI doesn't really mean anything as we're using extern to mean something slightly different to what it normally means. I think this is a point in favour of `repr(unsized)` but extern makes so much sense in that it heavily suggests what this feature is useful for.

Reply:
When it comes to bikeshedding, what about something like the following?
```rust
#[repr(opaque)]
struct Foo;
```

I personally don't like the use of `type` and I think `opaque` captures the meaning better than either `extern` or `unsized`. I also think it just makes more sense as a `repr`, and it would go nicely with specifying alignment (eg. `repr(opaque, align(16))`) if we allow that, which I think we should (though that could be added later if necessary).

Reply:
What's your reasoning for preferring `opaque` over `unsized`? I'm thinking of the common header use case where you may want:
```rust
#[repr(C, unsized)]
struct Header {
    field1: u8,
    field2: usize,
}
```

Reply:
These are described as opaque types throughout the RFC, which I think describes them well. The thing that makes them unique isn't that you don't know their size, but that you don't know (nor care about) their composition. This is in contrast to types like `[T]`, `str`, and `dyn Trait`, which are unsized, but not opaque.

Regarding the example you gave, if I didn't already know what it meant, I would find it very confusing. I wouldn't understand what `unsized` meant here, especially considering that the `struct` would appear to be composed entirely of sized types. It would be a lot less confusing, though, if the `struct` had no fields. 

To create a type with a non-opaque header followed by opaque data, I would imagine creating two separate types: one for the opaque data itself, and a separate type for the opaque data with a header (though that would seem to require `repr(align)` on the opaque type). Using the word `opaque` in this context (where the opaque type has no fields) would clearly indicate not only that we don't know the size, but that we don't know (nor care about) the composition of the type. Here's how I would imagine doing that:

```rust
#[repr(opaque, align(4)]
struct Opaque;

#[repr(C)]
struct OpaqueWithHeader {
    field1: u8,
    field2: usize,
    opaque: Opaque,
}
```

That said, I'm not opposed to either `extern type T` nor the `repr(unsized)` syntax. I just wanted to express my view that `repr(opaque)` might be a better fit. However, regardless of the syntax used, I would discourage the idea of allowing fields within the opaque type itself, as that would seem quite confusing to me.

Reply:
One issue I see with the `repr` attribute is that it looks too much like it's a unit struct declaration, if we're using a struct declaration I wonder if we could have native syntax for saying there are unknown fields (maybe with the `repr` too)

```rust
#[repr(opaque)]
struct Opaque {
  ..
}
```

Though, this has been proposed as syntax for `#[non_exhaustive]` previously. I think it fits opaqueness better since the fields are unknown even locally, whereas `#[non_exhaustive]` only affects metadata and the visible struct declaration _is_ what it is known to be locally.

Reply:
There's one place where `extern "C" { type X; }` and `extern "Rust" { type X; }` could differ: the former obviously should be FFI-safe, but the latter doesn't necessarily have to be. If and only if it's not, then it's presumably semver-compatible to replace it with a standard `struct` in the future.

Reply:
When I see "opaque (type)", I think of RPIT (return-position impl trait) types, and soon TAIT, RPITIT, etc.